File: src/1061.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   private:
    static const int maxUnionFind = 26;  // s1, s2, and baseStr consist of lowercase English letters.
    array<int, maxUnionFind> UnionFind;

    int find(int idx) {
        while (UnionFind[idx] != idx) {
            idx = UnionFind[idx];
        }

        return idx;
    }
    void join(int idx1, int idx2) {
        int root1 = find(idx1);
        int root2 = find(idx2);
        if (root1 == root2) {
            return;
        }

        if (root1 <= root2) {
            UnionFind[root2] = root1;
        } else {
            UnionFind[root1] = root2;
        }
    }

   public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        string retVal;

        for (long unsigned int i = 0; i < maxUnionFind; ++i) {
            UnionFind[i] = i;
        }

        for (long unsigned int i = 0; i < s1.length(); ++i) {
            join(s1[i] - 'a', s2[i] - 'a');
        }

        for (char c : baseStr) {
            retVal += (char)(find((c - 'a')) + 'a');
        }

        return retVal;
    }
};

int main(int argc, char **argv) {
    struct subject {
        string s1;
        string s2;
        string baseStr;
    };
    vector<subject> testData{
        {"parker", "morris", "parser"}, {"hello", "world", "hold"}, {"leetcode", "programs", "sourcecode"}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: s1 = "parker", s2 = "morris", baseStr = "parser"
     *  Output: "makkek"
     *
     *  Input: s1 = "hello", s2 = "world", baseStr = "hold"
     *  Output: "hdld"
     *
     *  Input: s1 = "leetcode", s2 = "programs", baseStr = "sourcecode"
     *  Output: "aauaaaaada"
     */

    Solution cSolution;
    string answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: s1 = \"" << testData[i].s1 << "\"";
        cout << ", s2 = \"" << testData[i].s2 << "\"";
        cout << ", baseStr = \"" << testData[i].baseStr << "\"\n";

        answer = cSolution.smallestEquivalentString(testData[i].s1, testData[i].s2, testData[i].baseStr);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/1061.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class UnionFind:
    def __init__(self, n) -> None:
        self.uf = list(range(n))

    def find(self, x) -> int:
        retVal = 0

        if self.uf[x] != x:
            self.uf[x] = self.find(self.uf[x])
        retVal = self.uf[x]

        return retVal

    def unite(self, x, y) -> None:
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return

        if x > y:
            x, y = y, x
        self.uf[y] = x


class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        retVal = ""

        uf = UnionFind(26)  # s1, s2, and baseStr consist of lowercase English letters.
        for a, b in zip(s1, s2):
            uf.unite(ord(a) - ord('a'), ord(b) - ord('a'))
        retVal = ''.join(chr(ord('a') + uf.find(ord(c) - ord('a'))) for c in baseStr)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for s1, s2, baseStr in zip(
                ["parker", "hello", "leetcode"],
                ["morris", "world", "programs"],
                ["parser", "hold", "sourcecode"]
        ):
            # /* Example
            #  *  Input: s1 = "parker", s2 = "morris", baseStr = "parser"
            #  *  Output: "makkek"
            #  *
            #  *  Input: s1 = "hello", s2 = "world", baseStr = "hold"
            #  *  Output: "hdld"
            #  *
            #  *  Input: s1 = "leetcode", s2 = "programs", baseStr = "sourcecode"
            #  *  Output: "aauaaaaada"
            #  */
            logging.info("Input: s1 = \"%s\", s2 = \"%s\", baseStr = \"%s\"", s1, s2, baseStr)

            retVal = pSolution.smallestEquivalentString(s1, s2, baseStr)
            logging.info("Output: \"%s\"", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
